apps/list.c

```diff
@@ -277,6 +279,73 @@ static void list_kdfs(void)
     sk_EVP_KDF_pop_free(kdfs, EVP_KDF_free);
 }
 
+/*
+ * Serialisers
+ */
+DEFINE_STACK_OF(OSSL_SERIALIZER)
+static int serialiser_cmp(const OSSL_SERIALIZER * const *a,
+                          const OSSL_SERIALIZER * const *b)
+{
+    int ret = OSSL_SERIALIZER_number(*a) - OSSL_SERIALIZER_number(*b);
+
+    if (ret == 0)
+        ret = strcmp(OSSL_PROVIDER_name(OSSL_SERIALIZER_provider(*a)),
+                     OSSL_PROVIDER_name(OSSL_SERIALIZER_provider(*b)));
+
+    return ret;
+}
+
+static void collect_serialisers(OSSL_SERIALIZER *serialiser, void *stack)
+{
+    STACK_OF(OSSL_SERIALIZER) *serialiser_stack = stack;
+
+    sk_OSSL_SERIALIZER_push(serialiser_stack, serialiser);
+    OSSL_SERIALIZER_up_ref(serialiser);
+}
+
+static void list_serialisers(void)
+{
+    STACK_OF(OSSL_SERIALIZER) *serialisers;
+    OSSL_PARAM params[3] = { OSSL_PARAM_END, OSSL_PARAM_END, OSSL_PARAM_END };
+    char *type, *format;
+    int i;
+
+    serialisers = sk_OSSL_SERIALIZER_new(serialiser_cmp);
+    BIO_printf(bio_out, "Provided SERIALISERs:\n");
+    OSSL_SERIALIZER_do_all_provided(NULL, collect_serialisers, serialisers);
+    sk_OSSL_SERIALIZER_sort(serialisers);
+    params[0] = OSSL_PARAM_construct_utf8_ptr(OSSL_SERIALIZER_PARAM_FORMAT,
+                                              &format, 0);
+    params[1] = OSSL_PARAM_construct_utf8_ptr(OSSL_SERIALIZER_PARAM_TYPE,
+                                              &type, 0);
+
+    for (i = 0; i < sk_OSSL_SERIALIZER_num(serialisers); i++) {
+        OSSL_SERIALIZER *k = sk_OSSL_SERIALIZER_value(serialisers, i);
+        STACK_OF(OPENSSL_CSTRING) *names =
+            sk_OPENSSL_CSTRING_new(name_cmp);
+
+        OSSL_SERIALIZER_names_do_all(k, collect_names, names);
+
+        format = type = NULL;
+
+        if (!OSSL_SERIALIZER_get_params(k, params))
+            format = type = "?";
+        BIO_printf(bio_out, "  ");
+        print_names(bio_out, names);
+        BIO_printf(bio_out, " @ %s for doing %s to %s serialisation\n",
+                   OSSL_PROVIDER_name(OSSL_SERIALIZER_provider(k)),
+                   type, format);
```

Comment:
I disagree with this output.  It would be much simpler, *and* more useful for programmers if you simply printed what `OSSL_SERIALIZER_properties()` returns.

Reply:
This simplified the code a lot.  Not quite so nice for the user.